File: ios/MorseWedaVideoEditorModule.swift
import ExpoModulesCore

// Expect the :app:shared Kotlin/Native framework to be available in the example app.
// On macOS, run ./gradlew :app:shared:assembleSharedReleaseFrameworkIosArm64 (or simulator)
// and embed the produced Shared.framework in the example app.
#if canImport(Shared)
import Shared
#endif

public class MorseWedaVideoEditorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MorseWedaVideoEditor")

    Events(
      "onExportProgress",
      "onExportComplete",
      "onExportError"
    )

    View(MorseWedaVideoEditorView.self) {
      Prop("licenseKey") { (view: MorseWedaVideoEditorView, key: String?) in
        #if canImport(Shared)
        if let key = key {
          LicenseGate.shared.setLicenseKey(key: key)
        }
        #endif
      }
    }

    Function("setLicenseCheckEnabled") { (enabled: Bool) in
      #if canImport(Shared)
      LicenseGate.shared.setLicenseCheckEnabled(enabled: enabled)
      #endif
    }

    Function("setLicenseKey") { (key: String) in
      #if canImport(Shared)
      LicenseGate.shared.setLicenseKey(key: key)
      #endif
    }

    AsyncFunction("initialize") { (options: [String: Any]) -> Void in
      #if canImport(Shared)
      let licenseKey = options["licenseKey"] as? String ?? ""
      let canCheckLicense = options["canCheckLicense"] as? Bool ?? false

      LicenseGate.shared.setLicenseCheckEnabled(enabled: canCheckLicense)
      if !licenseKey.isEmpty {
        LicenseGate.shared.setLicenseKey(key: licenseKey)
      }

      if canCheckLicense {
        let valid = LicenseGate.shared.checkLicense()
        if !valid {
          throw LicenseInvalidException()
        }
      }
      #else
      throw ModuleNotLinkedException()
      #endif
    }

    AsyncFunction("checkLicense") { () -> Bool in
      #if canImport(Shared)
      return LicenseGate.shared.checkLicense()
      #else
      return false
      #endif
    }

    AsyncFunction("exportVideo") { (timelineJson: String, outputPath: String, options: [String: Any]) -> String in
      #if canImport(Shared)
      let engine = ExportEngineKt.createExportEngine()
      let request = ExportRenderRequest(
        timelineJson: timelineJson,
        outputPath: outputPath,
        format: options["format"] as? String ?? "mp4",
        width: Int32((options["width"] as? NSNumber)?.intValue ?? 1280),
        height: Int32((options["height"] as? NSNumber)?.intValue ?? 720),
        fps: Int32((options["fps"] as? NSNumber)?.intValue ?? 30),
        bitrateMbps: Float((options["bitrateMbps"] as? NSNumber)?.floatValue ?? 8.0),
        isAudioMuted: (options["isAudioMuted"] as? Bool ?? false) ? 1 : 0,
        audioCodec: options["audioCodec"] as? String ?? "aac",
        audioBitrateKbps: Int32((options["audioBitrateKbps"] as? NSNumber)?.intValue ?? 128)
      )

      return try await withCheckedThrowingContinuation { continuation in
        let listener = ExportProgressListenerWrapper { current, total in
          self.sendEvent("onExportProgress", [
            "currentFrame": current,
            "totalFrames": total,
            "progress": total > 0 ? Float(current) / Float(total) : 0.0
          ])
        } onError: { code, message in
          self.sendEvent("onExportError", ["code": code, "message": message])
          continuation.resume(throwing: ExportException(message: message))
        } onComplete: { path in
          self.sendEvent("onExportComplete", ["outputPath": path])
          continuation.resume(returning: path)
        }

        guard let handle = engine.startExport(request: request, listener: listener), handle.isValid else {
          continuation.resume(throwing: ExportException(message: "Failed to start export"))
          return
        }
      }
      #else
      throw ModuleNotLinkedException()
      #endif
    }
  }
}

#if canImport(Shared)
class ExportProgressListenerWrapper: NSObject, ExportProgressListener {
  let onProgress: (Int32, Int32) -> Void
  let onError: (Int32, String) -> Void
  let onComplete: (String) -> Void

  init(onProgress: @escaping (Int32, Int32) -> Void, onError: @escaping (Int32, String) -> Void, onComplete: @escaping (String) -> Void) {
    self.onProgress = onProgress
    self.onError = onError
    self.onComplete = onComplete
  }

  func onProgress(currentFrame: Int32, totalFrames: Int32) {
    onProgress(currentFrame, totalFrames)
  }

  func onError(errorCode: Int32, errorMessage: String) {
    onError(errorCode, errorMessage)
  }

  func onComplete(outputPath: String) {
    onComplete(outputPath)
  }
}
#endif

class LicenseInvalidException: Exception {
  override var reason: String { "License validation failed" }
}

class ExportException: Exception {
  let message: String
  init(message: String) { self.message = message }
  override var reason: String { message }
}

class ModuleNotLinkedException: Exception {
  override var reason: String { "Morse Weda shared framework is not linked in this build" }
}

File: android/src/main/java/com/emmy/morsewedavideoeditor/MorseWedaVideoEditorModule.kt
package com.emmy.morsewedavideoeditor

import com.app.editor.LicenseGate
import com.app.editor.NativeVideoBridge
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.emmy.media_editor.src.video_editor.export.AndroidExportEngine
import org.emmy.media_editor.src.video_editor.export.ExportProgressListener
import org.emmy.media_editor.src.video_editor.export.ExportRenderRequest

/**
 * Expo module exposing the Morse Weda video editor to React Native.
 *
 * Route 1: drop-in UI via [MorseWedaVideoEditorView].
 * Route 2: headless hooks (initialize, license controls, exportVideo).
 */
class MorseWedaVideoEditorModule : Module() {

    private val ioScope = CoroutineScope(Dispatchers.IO)
    private var nativeBridge: NativeVideoBridge? = null
    private var exportEngine: AndroidExportEngine? = null

    override fun definition() = ModuleDefinition {
        Name("MorseWedaVideoEditor")

        Events(
            "onExportProgress",
            "onExportComplete",
            "onExportError"
        )

        View(MorseWedaVideoEditorView::class) {
            Prop("licenseKey") { view: MorseWedaVideoEditorView, key: String? ->
                key?.let { LicenseGate.setLicenseKey(it) }
            }
        }

        Function("setLicenseCheckEnabled") { enabled: Boolean ->
            LicenseGate.setLicenseCheckEnabled(enabled)
        }

        Function("setLicenseKey") { key: String ->
            LicenseGate.setLicenseKey(key)
        }

        AsyncFunction("initialize") { options: Map<String, Any?>, promise: Promise ->
            try {
                val licenseKey = options["licenseKey"] as? String ?: ""
                val canCheckLicense = options["canCheckLicense"] as? Boolean ?: false

                LicenseGate.setLicenseCheckEnabled(canCheckLicense)
                if (licenseKey.isNotEmpty()) {
                    LicenseGate.setLicenseKey(licenseKey)
                }

                nativeBridge = NativeVideoBridge()
                exportEngine = AndroidExportEngine()

                if (canCheckLicense) {
                    val valid = LicenseGate.checkLicense()
                    if (!valid) {
                        promise.reject(CodedException("LICENSE_INVALID", "License validation failed", null))
                        return@AsyncFunction
                    }
                }

                promise.resolve(null)
            } catch (e: Throwable) {
                promise.reject(CodedException("INIT_ERROR", e.message ?: "Unknown error", e))
            }
        }

        AsyncFunction("checkLicense") { promise: Promise ->
            val valid = LicenseGate.checkLicense()
            promise.resolve(valid)
        }

        AsyncFunction("exportVideo") { timelineJson: String, outputPath: String, options: Map<String, Any?>, promise: Promise ->
            val engine = exportEngine ?: run {
                promise.reject(CodedException("NOT_INITIALIZED", "Call initialize() first", null))
                return@AsyncFunction
            }

            val request = ExportRenderRequest(
                timelineJson = timelineJson,
                outputPath = outputPath,
                format = options["format"] as? String ?: "mp4",
                width = (options["width"] as? Number)?.toInt() ?: 1280,
                height = (options["height"] as? Number)?.toInt() ?: 720,
                fps = (options["fps"] as? Number)?.toInt() ?: 30,
                bitrateMbps = (options["bitrateMbps"] as? Number)?.toFloat() ?: 8f,
                isAudioMuted = options["isAudioMuted"] as? Boolean ?: false,
                audioCodec = options["audioCodec"] as? String ?: "aac",
                audioBitrateKbps = (options["audioBitrateKbps"] as? Number)?.toInt() ?: 128
            )

            ioScope.launch {
                try {
                    engine.startExport(request, object : ExportProgressListener {
                        override fun onProgress(currentFrame: Int, totalFrames: Int) {
                            appContext.mainQueue.launch {
                                this@MorseWedaVideoEditorModule.sendEvent("onExportProgress", mapOf(
                                    "currentFrame" to currentFrame,
                                    "totalFrames" to totalFrames,
                                    "progress" to if (totalFrames > 0) currentFrame.toFloat() / totalFrames else 0f
                                ))
                            }
                        }

                        override fun onError(errorCode: Int, errorMessage: String) {
                            appContext.mainQueue.launch {
                                this@MorseWedaVideoEditorModule.sendEvent("onExportError", mapOf(
                                    "code" to errorCode,
                                    "message" to errorMessage
                                ))
                                promise.reject(CodedException("EXPORT_ERROR", errorMessage, null))
                            }
                        }

                        override fun onComplete(outputPath: String) {
                            appContext.mainQueue.launch {
                                this@MorseWedaVideoEditorModule.sendEvent("onExportComplete", mapOf(
                                    "outputPath" to outputPath
                                ))
                                promise.resolve(outputPath)
                            }
                        }
                    })
                } catch (e: Throwable) {
                    appContext.mainQueue.launch {
                        promise.reject(CodedException("EXPORT_ERROR", e.message ?: "Export failed", e))
                    }
                }
            }
        }
    }
}
